src/html-imports.js

```diff
@@ -9,718 +9,754 @@
+    for (let i = n$.length - 1, n; i >= 0 && (n = n$[i]); i--) {
+      // Don't fire twice same event.
+      if (!n.__fired) {
+        n.__fired = true;
+        const eventType = n.import ? 'load' : 'error';
+        flags.log && console.warn('fire', eventType, n.href);
+        // Ensure the load promise is setup before firing the event.
+        whenElementLoaded(n);
+        n.dispatchEvent(new CustomEvent(eventType, {
+          bubbles: false,
+          cancelable: false,
+          detail: undefined
+        }));
       }
     }
   }
-  function loadedImport(e) {
-    markTargetLoaded(e);
-    newImports.push(this);
-    parsedCount++;
-    checkDone();
-  }
-  function errorLoadingImport(e) {
-    errorImports.push(this);
-    parsedCount++;
-    checkDone();
+
+  /**
+   * Waits for an element to finish loading. If already done loading, it will
+   * mark the element accordingly.
+   * @param {!Element} element
+   * @return {Promise}
+   */
+  function whenElementLoaded(element) {
+    if (!element.__loadPromise) {
+      element.__loadPromise = new Promise((resolve, reject) => {
+        if (isElementLoaded(element)) {
+          resolve(element);
+        } else {
+          element.addEventListener('load', () => resolve(element));
+          element.addEventListener('error', () => reject(element));
+        }
+      });
+    }
+    return element.__loadPromise;
   }
-  if (importCount) {
-    for (var i=0, imp; i<importCount && (imp=imports[i]); i++) {
-      if (isImportLoaded(imp)) {
-        newImports.push(this);
-        parsedCount++;
-        checkDone();
-      } else {
-        imp.addEventListener('load', loadedImport);
-        imp.addEventListener('error', errorLoadingImport);
+
+  /**
+   * @param {!Element} element
+   * @return {boolean}
+   */
+  function isElementLoaded(element) {
+    let isLoaded = false;
+    if (useNative && isImportLink(element) && element.import &&
+      element.import.readyState !== 'loading') {
+      isLoaded = true;
+    } else if (isIE && element.localName === 'style') {
+      // NOTE: IE does not fire "load" event for styles that have already
+      // loaded. This is in violation of the spec, so we try our hardest to
+      // work around it.
+      // If there's not @import in the textContent, assume it has loaded
+      if (element.textContent.indexOf('@import') == -1) {
+        isLoaded = true;
+        // if we have a sheet, we have been parsed
+      } else if (element.sheet) {
+        isLoaded = true;
+        const csr = element.sheet.cssRules;
+        // search the rules for @import's
+        for (let i = 0, l = csr ? csr.length : 0; i < l && isLoaded; i++) {
+          if (csr[i].type === CSSRule.IMPORT_RULE) {
+            // if every @import has resolved, fake the load
+            isLoaded = Boolean(csr[i].styleSheet);
+          }
+        }
       }
+    } else if (element.localName === 'script' && !element.src) {
```

Comment:
Won't this fail to notice that scripts and styles have loaded? I guess this is not required because of how it's used but this should probably be spelled out.